core/gpa_space.c

```diff
@@ -346,3 +357,118 @@ uint64 gpa_space_get_pfn(hax_gpa_space *gpa_space, uint64 gfn, uint8 *flags)
 
     return pfn;
 }
+
+int gpa_space_adjust_prot_bitmap(hax_gpa_space *gpa_space, uint64 max_gpfn)
+{
+    prot_bitmap *pb = &gpa_space->prot_bitmap;
+    uint8 *bmold = pb->bitmap, *bmnew = NULL;
+
+    /* Bitmap size only grows until it is destroyed */
+    if (max_gpfn <= pb->max_gpfn)
+        return 0;
+
+    bmnew = hax_vmalloc(gpa_space_prot_bitmap_size(max_gpfn), HAX_MEM_NONPAGE);
+    if (!bmnew) {
+        hax_error("%s: Not enought memory for new protection bitmap\n",
+                  __func__);
+        return -ENOMEM;
+    }
+    pb->bitmap = bmnew;
+    if (bmold) {
+        memcpy(bmnew, bmold, gpa_space_prot_bitmap_size(pb->max_gpfn));
+        hax_vfree(bmold, gpa_space_prot_bitmap_size(pb->max_gpfn));
+    }
+    pb->max_gpfn = max_gpfn;
+    return 0;
+}
+
+static void gpa_space_set_prot_bitmap(uint64 start, uint64 nbits,
+                                      uint8 *bitmap, bool set)
+{
+    uint64 i = 0;
+    uint64 start_index = start / 8;
+    uint64 start_bit = start % 8;
+    uint64 end_index = (start + nbits) / 8;
+    uint64 end_bit = (start + nbits) % 8;
+
+    if (start_index == end_index) {
+        for (i = start; i < start + nbits; i++)
+            if (set)
+                hax_test_and_set_bit(i, (uint64 *)bitmap);
+            else
+                hax_test_and_clear_bit(i, (uint64 *)bitmap);
+        return;
+    }
+
+    for (i = start; i < (start_index + 1) * 8; i++)
+        if (set)
+            hax_test_and_set_bit(i, (uint64 *)bitmap);
+        else
+            hax_test_and_clear_bit(i, (uint64 *)bitmap);
+
+    for (i = end_index * 8; i < start + nbits; i++)
+        if (set)
+            hax_test_and_set_bit(i, (uint64 *)bitmap);
+        else
+            hax_test_and_clear_bit(i, (uint64 *)bitmap);
+
+    for (i = start_index + 1; i < end_index; i++)
+        if (set)
+            bitmap[i] = 0xFF;
+        else
+            bitmap[i] = 0;
+}
+
+int gpa_space_test_prot_bitmap(struct hax_gpa_space *gpa_space, uint64 gfn)
+{
+    struct prot_bitmap *pbm = &gpa_space->prot_bitmap;
+
+    if (!pbm)
+        return 0;
+
+    if (gfn >= pbm->max_gpfn)
+        return 0;
+
+    return hax_test_bit(gfn, (uint64 *)pbm->bitmap);
+}
+
+int gpa_space_chunk_protected(struct hax_gpa_space *gpa_space, uint64 gfn,
```

Comment:
I see the QEMU patch make use of `HAX_CHUNK_SIZE` which is currently set to 2MB. When we introduced chunks with EPT2, we considered it an internal implementation detail and didn't expect to expose it to user space. If it's absolutely necessary, we should add an API to allow the user space to query the chunk size.
My question is, is making QEMU aware of HAXM chunks a necessity or an optimization? Admittedly, right now, at each non-MMIO EPT violation, we pin a whole chunk (instead of just the faulting page) in host RAM, and set up multiple EPT entries, so as to reduce the number of future EPT violations. But it's possible to make ept_tree_create_entries() skip the GFNs that are protected, so if the guest accesses one of them, we'll still get a VM exit. If we do that, is there any more reason for QEMU to know about chunks?

Reply:
In theory, (I mean I did not test it), QEMU does not need to know the chunk size. It can resolve only one page one time, or 1M, etc. It simply means in that case, QEMU would receive more EPT_FAULT_EXIT at the same place with different gpa reported. I did not design it in a way that user space has to rely on the exact chunk size.
For QEMU side patch, I have to agree I write it that way because I have knowledge on HAXM.
I think as long as driver does not make assumptions on user-space implementation, it should be good.

Reply:
So basically it's an optimization. I'd prefer that we process/fill one guest page per exit to QEMU, which is the same page that hardware (EPT violation) gave us, because

1. Having user space depend on a driver implementation detail is not a good practice, since the interface is not stable, although it's not as bad as having the driver depend on user space.
2. Chunks are created in HVA space (by dividing a RAM block), not GPA space. So technically, rounding a GPA down to the nearest 2MB boundary is not guaranteed to give you a GPA that maps to the same chunk.

Reply:
3. Some chunks are smaller than 2MB, e.g. in case of a tiny RAM block, or the last chunk in a RAM block, so `__gpa < gpa + 0x200000` doesn't always work.
4. HVF also fills one page at a time and takes this many VM exits. If we fill too many pages at a time, we may be doing kind of eager loading rather than on-demand loading (there's a trade-off).

If the performance is not good enough, we could add the chunk-based optimization in a later patch.